ACP/ThresholdMessageKit Work

derekpierre opened the pull request:
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [X] Other

**Required reviews:** 
- [ ] 1
- [ ] 2
- [X] 3

**What this does:**
> High-level idea of the changes introduced in this PR. 
> List relevant API changes (if any), as well as related PRs and issues.

**Issues fixed/closed:**
Depends on https://github.com/nucypher/ferveo/pull/156
Related to https://github.com/nucypher/nucypher/pull/3194

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
> What should reviewers focus on? 
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?

derekpierre committed:
Initial code and tests for AccessControlPolicy and its use by ThresholdDecryptionRequest.
Requires changes to DkgPublicKey in ferveo.

derekpierre committed:
Initial code for ThresholdMessageKit and tests.

derekpierre committed:
Add aad() function to AccessControlPolicy.

derekpierre committed:
Rename ThresholdMessageKit kem/dem ciphertext to header and payload respectively.

derekpierre committed:
Rename ThresholdMessageKit underlying member for the access control policy to "acp"; also rename accessor method.

derekpierre committed:
Add AccessControlPolicy and ThresholdMessageKit to `nucypher_core` python module.

derekpierre committed:
Update nucypher-core to incorporate latest changes from ferveo regarding use/exposure of key encapsulation i.e. use of CiphertextHader, and rework of Ciphertext.

fjarri commented on nucypher-core/src/versioning.rs:
```diff
@@ -136,6 +136,30 @@ pub trait ProtocolObjectInner<'a>: Serialize + Deserialize<'a> {
+    ///
+    /// NOTE: Underlying struct members for types that must remain backwards compatible should
+    /// also remain backwards compatible as well.
+    fn must_remain_backwards_compatible() -> (bool, u16) {
```
I understand that is the solution for smooth switching from development to the production version?

Not sure how I feel about it. Seems kind of brittle. Maybe just a different type would be the way to go. You'd have to do some additional work when switching to production, but that would solidify the commitment, so to speak. The new type may still save/check the last development major version in the serialized form, to allow the usage of objects created during development, but I'm not sure how much the convenience here outweighs the safety. Perhaps it is better to force the devs to start the production objects from a clean slate.

derekpierre replied:
Let's muse more about this in https://github.com/nucypher/nucypher-core/pull/75.

fjarri commented:
I think it might be better to extract the permanent-support objects into a separate PR, because it's pretty important and shouldn't be mixed up with the rest of the stuff that's going on here.

derekpierre commented:
> I think it might be better to extract the permanent-support objects into a separate PR, because it's pretty important and shouldn't be mixed up with the rest of the stuff that's going on here.

Removed the functionality from this PR and moved to https://github.com/nucypher/nucypher-core/pull/75.

derekpierre committed:
Modify AccessControlPolicy to be instantiated with an AuthenticatedData object which works around to chicken and egg issue for obtaining the aad to use when encrypting data, but which is needed for the AccessControlPolicy after the ciphertext is signed.
Update bindings/tests.

derekpierre committed:
Add encrypt_for_dkg function so that Ciphertext and AuthenticatedData can be returned from the same call.
Added python/wasm bindings.

derekpierre committed:
Make ThresholdMessageKit more efficient by not returning the Ciphertext object purely to get the CiphertextHeader.
Add decrypt_with_shared_secret to ThresholdMessageKit so that to decrypt data, the ciphertext does not need to be returned to python/js from Rust and then passed back in to Rust to decrypt. Keep it in Rust and just do the decryption internally.

derekpierre committed:
Update `nucypher-core` changelog with latest changes.

derekpierre committed:
Fix mypy build error. Related to this https://pyo3.rs/main/function/signature#trailing-optional-arguments.

piotr-roslaniec commented on nucypher-core-python/nucypher_core/__init__.pyi:
```diff
@@ -429,22 +431,89 @@ class MetadataResponse:
+@final
+class AuthenticatedData:
+
+    def __init__(self, public_key: DkgPublicKey, conditions: Optional[Conditions]):
```
@derekpierre I just recalled your comment on `Optional<&Condition>` in DKG structures. Are conditions supposed to be optional here too?

derekpierre replied:
It is downstream.

This is something that came to me last night, but it pervades many types in `nucypher-core` (ThresholdDecryptionRequest, ... etc.) and I didn't have time to fix it across all of them. I think we should fix it in a separate PR. Basically for DKG types Conditions should not be Optional but should be specified. Hence I filed #76 .

piotr-roslaniec commented on nucypher-core-wasm/src/lib.rs:
```diff
@@ -648,6 +653,167 @@ impl SessionSecretFactory {
+        self.0.conditions.clone().map(Conditions)
+    }
+
+    #[wasm_bindgen(js_name = toBytes)]
+    pub fn to_bytes(&self) -> Box<[u8]> {
+        to_bytes(self)
+    }
```
We could use `generate_to_bytes` here

piotr-roslaniec commented on nucypher-core-wasm/src/lib.rs:
```diff
@@ -648,6 +653,167 @@ impl SessionSecretFactory {
+        self.0.auth_data.conditions.clone().map(Conditions)
+    }
+
+    #[wasm_bindgen(js_name = toBytes)]
+    pub fn to_bytes(&self) -> Box<[u8]> {
+        to_bytes(self)
+    }
```
We could use `generate_to_bytes` here

derekpierre replied:
✔️

piotr-roslaniec commented on nucypher-core-wasm/src/lib.rs:
```diff
@@ -648,6 +653,167 @@ impl SessionSecretFactory {
+//
+#[wasm_bindgen]
+#[derive(PartialEq, Eq, Debug, derive_more::From, derive_more::AsRef)]
+pub struct ThresholdMessageKit(nucypher_core::ThresholdMessageKit);
```
I've noticed that in some places in `nucypher-core`, we're using `{ backend: InnerType }` and `(InnerType)` in others. Perhaps we could settle on one way to define structs. It's not relevant to this PR, just a general comment on style

piotr-roslaniec commented on nucypher-core-wasm/src/lib.rs:
```diff
@@ -648,6 +653,167 @@ impl SessionSecretFactory {
+            .map_err(map_js_err)
+    }
+
+    #[wasm_bindgen(js_name = toBytes)]
+    pub fn to_bytes(&self) -> Box<[u8]> {
+        to_bytes(self)
+    }
```
We could use `generate_to_bytes` here

derekpierre replied:
✔️

piotr-roslaniec commented on nucypher-core/src/access_control.rs:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    /// Return the aad.
+    pub fn aad(&self) -> Box<[u8]> {
+        let public_key_bytes = self.public_key.to_bytes().unwrap();
+        let condition_bytes = self.conditions.as_ref().unwrap().as_ref().as_bytes();
+        let mut result = Vec::with_capacity(public_key_bytes.len() + condition_bytes.len());
+        result.extend(public_key_bytes);
+        result.extend(condition_bytes);
+        result.into_boxed_slice()
+    }
```
I'm tempted to suggest this functional approach
```suggestion
	pub fn aad(&self) -> Box<[u8]> {
	    [self.public_key.to_bytes(), self.conditions.as_ref().as_ref().as_bytes()]
	        .concat()
	        .into_boxed_slice()
	}
```

derekpierre replied:
@piotr-roslaniec Can you take care of this change for me - either as part of this PR or a follow-up 🙏 ?

piotr-roslaniec replied:
Sure, we can address any non-critical changes in a follow-up PR. I can take care of that once I test this PR with `nucypher-ts`.

piotr-roslaniec commented on nucypher-core/src/access_control.rs:
```diff
@@ -0,0 +1,209 @@
+    }
+}
+
+impl Eq for AuthenticatedData {}
```
Can we `#[derive(Eq)]` on `AuthenticatedData` instead? I assume there is a reason deriving `Eq` in this particular way.

derekpierre replied:
I may have mistakenly thought that you have to do this if you implemented `PartialEq`. Was that mistaken? In any case, it's not a blocker right...?

piotr-roslaniec replied:
While working on #77 I realized that automatic derivation of `Eq` is not possible since `DkgPublicKey` from `ferveo` doesn't implement it either. 

So it makes sense to provide an implementation stub `impl Eq {}` as a workaround. We need `Eq` do implement `ProtocolObjectInner` for `AccessControlPolicy`. Tracking this here: https://github.com/nucypher/ferveo/issues/157

piotr-roslaniec commented on nucypher-core/src/threshold_message_kit.rs:
```diff
@@ -0,0 +1,109 @@
+    messagepack_deserialize, messagepack_serialize, ProtocolObject, ProtocolObjectInner,
+};
+
+// TODO should this be in umbral?
```
@derekpierre After our chat about `encrypt_for_dkg` yesterday I have considered pulling some of the related code into `ferveo`. So a similar notion crossed my mind. However, since neither of those are directly related to cryptographic concept in `umbral` or `ferveo`, I think these are best saved for `nucypher-core` where other protocol-specific concepts reside.

derekpierre replied:
This is comment that gets moved by cargo fmt.

It was meant for serde_bytes, which is in umbral, but probably should be in a more common place.

derekpierre replied:
This is a comment that gets moved by the linter.

It was meant for serde_bytes (used previously), that we use in `nucypher-core` but it is implemented in `umbral`. Feels like that should be in some common library, or moved into serde itself. I believe Bogdan was trying to get it in serde.

piotr-roslaniec replied:
Ah, I see. I also experienced these antics from `cargo fmt`

piotr-roslaniec reviewed (commented):
Looks like a solid piece of work.

I have some comments, mostly nitpicks. AFAIK this is functionally correct. 

I'm not approving yet because I want to test these changes in `nucypher-ts`.

derekpierre committed:
Use to_bytes macro for dkg wasm bindings.

derekpierre committed:
Use equals from DkgPublicKey directly for AuthenticatedData equals.

derekpierre commented on nucypher-core/src/access_control.rs:
```diff
@@ -0,0 +1,209 @@
+impl PartialEq for AuthenticatedData {
+    fn eq(&self, other: &Self) -> bool {
+        self.public_key.to_bytes().unwrap() == other.public_key.to_bytes().unwrap()
+            && self.conditions == other.conditions
```
Fixed. I think (?) DkgPublicKey from ferveo didn't previously use PartialEq but it seems now it does in a newer version?

In any case - fixed it ✔️

jMyles reviewed (commented):
Just did a first read.  Looks good on first read.  :-)

piotr-roslaniec reviewed (approved):
Successfully tested in https://github.com/nucypher/nucypher-ts/pull/273
